### PR TITLE
Consider hashing resources to derive delta version number instead of passing a version directly

### PR DESCRIPTION
**Pull request by mikegajda**
This [PR](https://github.com/envoyproxy/go-control-plane/pull/387) in the Go repo had me thinking about how versions were generated at the resource level. The Go PR [hashes](https://github.com/envoyproxy/go-control-plane/pull/387/files#diff-9ac26beeceb864dff60e72228fd6421bad7b7a29bdcc8df90280b3ee7e54c7a4R163-R176) the resource and uses that as the version. I think this is a solid approach because it would allow for a non-breaking change for merging these delta changes.

Originally `SnapshotResource` was introduced to wrap a resource with a `version` String so that DeltaRequests could send updates only for resources whose version had changed. This PR retains `SnapshotResource` as we need a binding of resource -> version for the delta xds bits, but no longer requires a `version` to be passed in. This allows users to pass in an `Iterable` of the resource (like before).

This version generation will impose some costs on all users as each resource will need to be hashed. 

TODO
- [x] check about the stability of the hash that relies on `resource.toString()` as noted in comments
- [ ] think about a way to only hash when running in delta mode

### CODE REVIEW

File: server/src/test/java/io/envoyproxy/controlplane/server/V3DiscoveryServerAdsDeltaResourcesIT.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        LISTENER_PORT,
+        "route0",
+        "2");
+    LOGGER.info("snapshot={}", snapshot);
+    cache.setSnapshot(
+        GROUP,
+        snapshot
+    );
+
+    TimeUnit.SECONDS.sleep(3);
```
Comment (slonka) [inline server/src/test/java/io/envoyproxy/controlplane/server/V3DiscoveryServerAdsDeltaResourcesIT.java:127]: To reduce test time please use awaitility.

File: server/src/test/resources/envoy/launch_envoy.sh [hunk 2]
```diff
@@ -13,6 +13,6 @@ echo "${CONFIG}" | sed -e "s/HOST_IP/${HOST_IP}/g" -e "s/HOST_PORT/${HOST_PORT}/
 cat "${CONFIG_FILE}"
 
 shift 2
-/usr/local/bin/envoy --drain-time-s 1 -c "${CONFIG_FILE}" "$@"
+/usr/local/bin/envoy -c "${CONFIG_FILE}" "$@"
```
Comment (slonka) [inline server/src/test/resources/envoy/launch_envoy.sh:16]: Why did you remove the drain time?

File: server/src/test/java/io/envoyproxy/controlplane/server/DockerContainerTest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+import io.restassured.http.ContentType;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.Network;
+
+public class DockerContainerTest {
```
Comment (slonka) [inline server/src/test/java/io/envoyproxy/controlplane/server/DockerContainerTest.java:14]: Not sure what's the role of this test.

**Review by slonka — COMMENTED** [review]
I will need a second round of review to make sure I covered all the cases.

File: server/src/test/java/io/envoyproxy/controlplane/server/DockerContainerTest.java (hunk 3 above)
Comment (mikegajda) [reply server/src/test/java/io/envoyproxy/controlplane/server/DockerContainerTest.java:14]: Ah, this is leftover from a dev cycle, deleting

File: server/src/test/java/io/envoyproxy/controlplane/server/EnvoyContainer.java [hunk 4]
```diff
@@ -41,10 +42,9 @@ protected void configure() {
     withCommand(
         "/bin/sh", "/usr/local/bin/launch_envoy.sh",
         Integer.toString(controlPlanePortSupplier.get()),
-        CONFIG_DEST,
-        "-l", "debug");
+        CONFIG_DEST
```
[the remark this answers is not in the document]
Comment (mikegajda) [reply server/src/test/java/io/envoyproxy/controlplane/server/EnvoyContainer.java:45]: Adding back

File: server/src/test/resources/envoy/launch_envoy.sh (hunk 2 above)
Comment (mikegajda) [reply server/src/test/resources/envoy/launch_envoy.sh:16]: This is also something I forgot to put back in, thought it was problematic with envoy v1.17/18, but turned out not to be the case

File: cache/src/main/java/io/envoyproxy/controlplane/cache/SnapshotResource.java [hunk 5]
```diff
@@ -20,6 +22,22 @@
...
+  /**
+   * Returns a new {@link SnapshotResource} instance.
+   *
+   * @param resource the resource
+   * @param <T>      the type of resource
+   */
+  public static <T extends Message> SnapshotResource<T> create(T resource) {
+    return new AutoValue_SnapshotResource<>(
+        resource,
+        // todo: is this a stable hash?
```
[the remark this answers is not in the document]
Comment (mikegajda) [reply cache/src/main/java/io/envoyproxy/controlplane/cache/SnapshotResource.java:34]: Should have been clearer here, I'm more interested in checking that `resource.toString()` doesn't obscure any information that could result in two distinct resources having a hash collision

File: server/src/test/java/io/envoyproxy/controlplane/server/V3DiscoveryServerAdsDeltaResourcesIT.java (hunk 1 above)
Comment (mikegajda) [reply server/src/test/java/io/envoyproxy/controlplane/server/V3DiscoveryServerAdsDeltaResourcesIT.java:127]: done

File: cache/src/main/java/io/envoyproxy/controlplane/cache/SnapshotResource.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (slonka) [reply cache/src/main/java/io/envoyproxy/controlplane/cache/SnapshotResource.java:34]: toString() calls `getAllFields` (and also prints unknownFields) so we should be good here.

**Comment by slonka** [comment]
Thanks for more tests, I'll take a look at it this week.

**Comment by mikegajda** [comment]
Awesome, looking forward to your feedback

File: server/src/test/java/io/envoyproxy/controlplane/server/V3DeltaDiscoveryServerCallbacks.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  @Override
+  public void onV2StreamRequest(long streamId,
+      io.envoyproxy.envoy.api.v2.DiscoveryRequest request) {
+    throw new IllegalStateException("Unexpected v2 request in v3 test");
+  }
+
+  @Override
+  public void onV3StreamRequest(long streamId, DiscoveryRequest request) {
+    LOGGER.error("request={}", request);
```
Comment (slonka) [inline server/src/test/java/io/envoyproxy/controlplane/server/V3DeltaDiscoveryServerCallbacks.java:57]: I think it's a leftover from a debugging session right?

Comment (slonka) [inline server/src/test/java/io/envoyproxy/controlplane/server/V3DeltaDiscoveryServerCallbacks.java:57]: Now I know it's from the tests package but we should probably keep the log levels the same.

**Review by slonka — COMMENTED** [review]
Overall looks fine, I think we should probably invest some time in DRY-ing up the tests, but this is something unrelated to this change. I will be testing out this feature in https://github.com/allegro/envoy-control/ and if nothing worrying pops up I think we should move forward with this PR in the main repo.

**Comment by mikegajda** [comment]
Also looking into running with the Allegro tests, noticing some breaking changes, looking into if we can make them not breaking to avoid having to make changes to Allegro to get this to run

**Comment by mikegajda** [comment]
Hey @slonka checking in here. Tried to get around the breaking changes, but got pulled into other things the past week. Is that a path that would be useful for your testing (getting this to be a non-breaking change) or are you already past that?

**Comment by slonka** [comment]
Hi,

I had some time off so I haven't started yet. If you have time and you can
implement it without breaking changes that would be awesome.

wt., 30 mar 2021, 15:05 użytkownik Mike Gajda ***@***.***>
napisał:

**Comment by mikegajda** [comment]
Closing this in favor of: https://github.com/HubSpot/java-control-plane/pull/4
